CSV Center Splitter Gear

echeng06 opened the pull request:
Adds the CSV Center Splitter gear, which splits an input CSV by an ADCID header key and writes results to the corresponding target project.

Changes to common code/other gears:
* Extracts the `build_project_map` method from the `pull_metadata` gear into a common `projects/project_mapper.py`
    * Also adds functionality to only act on a subset of centers 
* Adds a `split_csv_by_key` method to `inputs/csv_reader.py`
* Adds a `write_csv_to_project` method to `outputs/outputs.py`
* Adds cookiecutter code for generating a gear's `docs`, and also updates/cleans up some of the older templating

bjkeller reviewed (commented):
I've asked for a couple of changes. We definitely should move build_project_mapper away from using the adcid tags.

Also, can you run `pants check gear/csv_center_splitter::` and fix whatever comes up there?  some of these may just require some overrides, for instance InputFileWrapper.create will return None if the input file is optional, which can be ignored if the input file is required

echeng06 commented on gear/csv_center_splitter/src/docker/manifest.json:
```diff
@@ -0,0 +1,76 @@
+            "description": "Column name from the input CSV with the ADCIDs",
+            "type": "string"
+        },
+        "target_project": {
```
Project name, since it can be for multiple centers. I can update the description to clarify. `staging_project_id` is the one where it's an ID and staged at a single project and is intended to override `target_project` if specified.

echeng06 commented on common/src/python/outputs/outputs.py:
```diff
@@ -72,3 +76,32 @@ def object_list(self) -> List[Dict[str, Any]]:
+    for row in data:
+        writer.write(row)
+
+    contents = stream.getvalue()
```
The way different gears define the FileSpec for their uses isn't really generic without needing to pass all the fields to pass to FileSpec anyways - maybe in this case it makes more sense to just define the FileSpec out in this gear's `main.py` like the other gears are doing?

echeng06 replied:
Also would it be okay to just return the StringIO object - I can see it being useful having the entire StringIO object instead (there is actually a test I wrote in a different gear that would need something like that haha).

bjkeller replied:
I would go with consistency. Though I'm pushing toward hiding FW specifics, which means either hiding them in functions/classes or pulling them out at least to main.py

Returning the StringIO object is fine

echeng06 commented:
@bjkeller Thanks for the review! I've updated as much as I could but just as an FYI I want to incorporate the changes in this PR before testing/finalizing all the way: https://github.com/naccdata/flywheel-gear-extensions/pull/128

echeng06 commented:
@bjkeller Alrighty I believe I've addressed the comments, apologies for how messy this got due to all the linting/formatting fixes, please take another look whenever you have time, thanks!

bjkeller reviewed (changes requested):
Thanks for the changes, but please let me know about that exception, and let's remove the merged cells option.

Also can you look at the type errors from 
`pants check common/test/python/centers::` and
`pants check gear/center_management::`?

Also, I think just let yapf do what it wants when you run `pants fix ::`. I don't always agree with it but it's not worth fighting it

echeng06 commented on gear/csv_center_splitter/src/docker/manifest.json:
```diff
@@ -0,0 +1,81 @@
+            "type": "string",
+            "default": ""
+        },
+        "allow_merged_cells": {
```
Ah okay - I added it because some of the NCRAD files have merged cells, which end up looking like this when converted to CSV:

<img width="1378" alt="image" src="https://github.com/user-attachments/assets/a215add9-1437-494e-a4cd-34bb76480552" />

I think there was a conversation saying it was fine to leave them like that but I could be misremembering/we can ask NCRAD to fill in the blanks. Either way I'll remove it.

bjkeller replied:
they aren't supposed to be doing that

echeng06 commented:
Okay I removed the merged cells thing, I have to switch gears (no pun intended) back to UDSv4 for a bit but I'll fix the linting/formatting stuff after!

bjkeller reviewed (approved):
Ready to merge. Thanks!
